### PR TITLE
taker checks for duplicate utxos now in recv_txio()

### PR DESCRIPTION
**Pull request by chris-belcher**
Please review

### CODE REVIEW

File: joinmarket/taker.py [hunk 1]
```diff
@@ -112,14 +112,19 @@ def recv_txio(self, nick, utxo_list, cj_pub, change_addr):
             log.debug(('recv_txio => nick={} not in '
                        'nonrespondants {}').format(nick, self.nonrespondants))
             return
-        self.utxos[nick] = utxo_list
-        utxo_data = jm_single().bc_interface.query_utxo_set(self.utxos[nick])
+        duplicate_utxos = set(sum(self.utxos.values(), [])).intersection(set(utxo_list))
```
Comment (adlai) [inline joinmarket/taker.py]: - `self.utxos.values()` is the sets of coins sent by each nick as inputs
- `sum(list_of_lists, [])` is an idiom for [flatten](https://stackoverflow.com/questions/406121/flattening-a-shallow-list-in-python#comment1198171_408281)

File: joinmarket/taker.py [hunk 2]
```diff
@@ -112,14 +112,19 @@ def recv_txio(self, nick, utxo_list, cj_pub, change_addr):
...
+        if len(duplicate_utxos) > 0:
+            log.debug('ERROR duplicate UTXOs: \n' + pprint.pformat(
+                duplicate_utxos))
+            return
```
Comment (adlai) [inline joinmarket/taker.py]: - maker providing duplicate coin effectively treated as non responding

File: joinmarket/taker.py (hunk 1 above)
Comment (chris-belcher) [inline joinmarket/taker.py]: effectively [] + list_of_lists[0] + list_of_lists[1] + ...

**Comment by AdamISZ** [comment]
Although the taker could choose to infer that *both* copies are from a "naughty" maker and treat both as non-responding, I presume there (might?) be a dos vector that way (although a bit theoretical? auth is under encryption).

Do we consider that this needs a test that triggers the error?

**Comment by chris-belcher** [comment]
Yes you're right, it could blame bother makers rather than the one that happened to respond second. That should've be too hard, I'll add it to that branch soonish.

Yes a test would be great. I'll add that too (maybe ask you guys on IRC some details about it)

**Comment by chris-belcher** [comment]
The reason I've left this so long is it's not trivial, there would need to be some records kept in the code for which maker contributed what in fees and utxos, so that both makers can be evicted if they clash.

**Pull request closed without merging**